src/base/profile.cpp

```diff
@@ -33,28 +33,29 @@
 
 Profile *Profile::m_instance = nullptr;
 
-Profile::Profile(Private::Profile *impl, Private::PathConverter *pathConverter)
-    : m_profileImpl(impl)
-    , m_pathConverterImpl(pathConverter)
+Profile::Profile(const QString &rootProfilePath, const QString &configurationName, const bool convertPathsToProfileRelative)
 {
+    if (rootProfilePath.isEmpty())
+        m_profileImpl = std::make_unique<Private::DefaultProfile>(configurationName);
+    else
+        m_profileImpl = std::make_unique<Private::CustomProfile>(rootProfilePath, configurationName);
+
     ensureDirectoryExists(SpecialFolder::Cache);
     ensureDirectoryExists(SpecialFolder::Config);
     ensureDirectoryExists(SpecialFolder::Data);
+
+    if (convertPathsToProfileRelative)
+        m_pathConverterImpl = std::make_unique<Private::Converter>(m_profileImpl->baseDirectory());
+    else
+        m_pathConverterImpl = std::make_unique<Private::NoConvertConverter>();
 }
 
 void Profile::initInstance(const QString &rootProfilePath, const QString &configurationName,
     const bool convertPathsToProfileRelative)
 {
     if (m_instance)
         return;
-
-    std::unique_ptr<Private::Profile> profile(rootProfilePath.isEmpty()
```

Comment:
Could you briefly explain why is it necessary to move these lines into `Profile` constructor, it is unclear to me why it wouldn't work here.

Reply:
Because we need `profile->baseDirectory()` at https://github.com/qbittorrent/qBittorrent/blob/316f34cbf5e956cd1381d85aff30f9eafd4c4714/src/base/profile.cpp#L55 to return already created directory by https://github.com/qbittorrent/qBittorrent/blob/316f34cbf5e956cd1381d85aff30f9eafd4c4714/src/base/profile.cpp#L40-L42 or it will return empty string if directory doesn't exist yet (when running in portable mode for the first time with empty `profile` directory).

Talked about it already [here](https://github.com/qbittorrent/qBittorrent/pull/11785#pullrequestreview-336977740) and [here](https://github.com/qbittorrent/qBittorrent/pull/11785#pullrequestreview-342287725).

Reply:
Understood.